include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -157,28 +160,94 @@ struct FFT<2,true, Scalar, device_capabilities>
         }
-        FFT<2,Inverse,device_capabilities>::template __call<Accessor,SharedMemoryAccessor>(access,sharedMemAccessor);
-        if (Inverse)
+
+        // do K/2 small workgroup FFTs
+        DynamicOffsetAccessor < Accessor, complex_t<Scalar> > offsetAccessor;
+        //[unroll(K/2)]
+        for (uint32_t k = 0; k < K; k += 2)
         {
-           ... special steps ...
+            if (k)
+            sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+            offsetAccessor.offset = _NBL_HLSL_WORKGROUP_SIZE_*k;
+            FFT<2,false, Scalar, device_capabilities>::template __call(offsetAccessor,sharedmemAccessor);
         }
+        accessor = offsetAccessor.accessor;
     }
 };
 
-*/
+// Inverse FFT
+template<uint32_t K, typename Scalar, class device_capabilities>
+struct FFT<K, true, Scalar, device_capabilities>
+{
+    template<typename Accessor, typename SharedMemoryAccessor>
+    static enable_if_t< (mpl::is_pot_v<K> && K > 2), void > __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor)
+    {
+        // do K/2 small workgroup FFTs
+        DynamicOffsetAccessor < Accessor, complex_t<Scalar> > offsetAccessor;
+        //[unroll(K/2)]
+        for (uint32_t k = 0; k < K; k += 2)
+        {
+            if (k)
+            sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+            offsetAccessor.offset = _NBL_HLSL_WORKGROUP_SIZE_*k;
+            FFT<2,true, Scalar, device_capabilities>::template __call(offsetAccessor,sharedmemAccessor);
+        }
+        accessor = offsetAccessor.accessor;
+      
+        for (uint32_t stride = _NBL_HLSL_WORKGROUP_SIZE_ << 1; stride <= (K >> 1) * _NBL_HLSL_WORKGROUP_SIZE_; stride <<= 1)
+        {
+            //[unroll(K/2)]
+            for (uint32_t virtualThreadID = SubgroupContiguousIndex(); virtualThreadID < (K >> 1) * _NBL_HLSL_WORKGROUP_SIZE_; virtualThreadID += _NBL_HLSL_WORKGROUP_SIZE_)
+            {
+                const uint32_t loIx = ((virtualThreadID & (~(stride - 1))) << 1) | (virtualThreadID & (stride - 1));
+                const uint32_t hiIx = loIx | stride;
+                
+                complex_t<Scalar> lo = accessor.get(loIx);
+                complex_t<Scalar> hi = accessor.get(hiIx);
+
+                hlsl::fft::DIT<Scalar>::radix2(hlsl::fft::twiddle<true,Scalar>(virtualThreadID & (stride - 1), stride), lo,hi);
+                
+                // Divide by special factor at the end
+                if ( (K >> 1) * _NBL_HLSL_WORKGROUP_SIZE_ == stride)
+                {
+                    divides_assign< complex_t<Scalar> > divAss;
+                    divAss(lo, K >> 1);
+                    divAss(hi, K >> 1);  
+                }
+
+                accessor.set(loIx, lo);
+                accessor.set(hiIx, hi);
+            }
+            accessor.memoryBarrier(); // no execution barrier just making sure writes propagate to accessor
```

Comment:
you don't need a memory barrier after the last iteration of this.

But you need a memory barrier between the first iteration of this loop and the mini-FFTs you did before.

Therefore more the barrier to the start of the scope (before the unrolled loop)